retester: Added config file (#268)
openshift/ci-docs: content/en/docs/architecture/retester.md

Retester is aware of Prow and the concept of optional and required jobs. It only triggers retests on PRs 
where at least one required Prow job is failing.

We can enable the retester on a repository or for the entire organization in [its configuration file](https://github.com/openshift/release/blob/master/core-services/retester/_config.yaml) and exclude it from [the old retester](https://github.com/openshift/release/blob/05dd9a1ab5881e55165a0cc0f40d5513e2e2fd11/ci-operator/jobs/infra-periodics.yaml#L260-L300) by `-repo:<org>/<repo>` in the `query`. 

### Config file
There are 3 levels of policy in the configuration file: retester (global), org, repo. If a field of policy is defined at two levels, the value at the more specific level precedes. The values in the more generic level are inherited if they are not defined in the more specific level.

```yaml
`enabled`:
  - `true`: enable retester
  - `false`: disable retester
  - not defined: use the value in the above level
`max_retests_for_sha_and_base`: the number of retesting for each sha and base
`max_retests_for_sha`: the number of retesting for each sha
```

> When merging policies, a 0 value at `max_retests` results in inheriting the parent policy.

Reviewer: I missed this in the last time.

```suggestion
```

> ### Back-off

Reviewer: ```suggestion
```

nit: I would remove the section title because it is part of the configuration file now.

> Retests are paused after three attempts against one base/PR HEAD combination, and the PR is explicitly held (`/hold`) after nine retests of a single PR HEAD.

Reviewer: ```suggestion
Retests are paused after `max_retests_for_sha_and_base` attempts against one base/PR HEAD combination, and the PR is explicitly held (`/hold`) after `max_retests_for_sha` retests of a single PR HEAD.
```

Retester is aware of Prow and the concept of optional and required jobs. It only triggers retests on PRs 
where at least one required Prow job is failing.

> We can enable the retester on a repository or for the entire organization in configuration file. Configuration file is added by adding a flag `--config-file=<file>` to [the retester's deployment](https://github.com/openshift/release/blob/master/clusters/app.ci/prow/03_deployment/retester.yaml) and exclude it from [the old retester](https://github.com/openshift/release/blob/05dd9a1ab5881e55165a0cc0f40d5513e2e2fd11/ci-operator/jobs/infra-periodics.yaml#L260-L300) by `-repo:<org>/<repo>` in the `query`. The following example shows how to achieve it for repo `openshift/ci-tools`:

Reviewer 2: ```suggestion
We can enable the retester on a repository or for the entire organization in a configuration file. Configuration file is added by adding a flag `--config-file=<file>` to [the retester's deployment](https://github.com/openshift/release/blob/master/clusters/app.ci/prow/03_deployment/retester.yaml) and exclude it from [the old retester](https://github.com/openshift/release/blob/05dd9a1ab5881e55165a0cc0f40d5513e2e2fd11/ci-operator/jobs/infra-periodics.yaml#L260-L300) by `-repo:<org>/<repo>` in the `query`. The following example shows how to achieve it for repo `openshift/ci-tools`:
```

> We can enable the retester on a repository or for the entire organization in a configuration file. Configuration file is added by adding a flag `--config-file=<file>` to [the retester's deployment](https://github.com/openshift/release/blob/master/clusters/app.ci/prow/03_deployment/retester.yaml) and exclude it from [the old retester](https://github.com/openshift/release/blob/05dd9a1ab5881e55165a0cc0f40d5513e2e2fd11/ci-operator/jobs/infra-periodics.yaml#L260-L300) by `-repo:<org>/<repo>` in the `query`. The following example shows how to achieve it for repo `openshift/ci-tools`:

Reviewer: ```suggestion
We can enable the retester on a repository or for the entire organization in [its configuration file](https://github.com/openshift/release/blob/master/core-services/retester/_config.yaml) and exclude it from [the old retester](https://github.com/openshift/release/blob/05dd9a1ab5881e55165a0cc0f40d5513e2e2fd11/ci-operator/jobs/infra-periodics.yaml#L260-L300) by `-repo:<org>/<repo>` in the `query`.

Reviewer: With the link to the configuration file, we can remove the two yaml examples.

        -repo:openshift/ci-tools                  # Exclude repo openshift/ci-tools
{{< / highlight >}}

### Config file
There are 3 levels of policy in the configuration file retester/org/repo. The policy has 3 attributes: `enabled`, `max_retests_for_sha`, `max_retests_for_sha_and_base`.
When merging policies, a 0 value at `max_retests` results in inheriting the parent policy.

For attribute `enabled` these rules apply:
- In level repo `enabled: false` means disabled repo. Nothing can change that.
- In level org `enabled: true` means enabled org and all its repos. But repo itself can be disabled.
- In level org `enabled: false` means disabled org and all its repos. But repo itself can be enabled.

> - In level retster `enabled` doesn't do anything.

Reviewer 2: ```suggestion
- In level retester `enabled` doesn't do anything.
```

        -repo:openshift/ci-tools                  # Exclude repo openshift/ci-tools
{{< / highlight >}}

### Config file

> There are 3 levels of policy in the configuration file: retester (global), org, repo. The policy has 3 attributes: `enabled`, `max_retests_for_sha`, `max_retests_for_sha_and_base`.

Reviewer: ```suggestion
There are 3 levels of policy in the configuration file: retester (global), org, repo. If a field of policy is defined at two levels, the value at the more specific level precedes. The values in the more generic level are inherited if they are not defined in the more specific level.

```yaml
`enabled`:
  - `true`: enable retester
  - `false`: disable retester
  - not defined: use the value in the above level
`max_retests_for_sha_and_base`: the number of retesting for each sha and base
`max_retests_for_sha`: the number of retesting for each sha
```

```

Reviewer: Should be OK without the paragraphs below.
